### PR TITLE
TY-2270 Feedback loop from Dart

### PR DESCRIPTION
**Pull request by biernacki**
[Jira ref - DocumentFeedbackChanged](https://xainag.atlassian.net/browse/TY-2311)
[Jira ref - DocumentTimeSpent](https://xainag.atlassian.net/browse/TY-2271)

- added `timeLogged` and `userReacted` to the `Engine` abstract class to communicate with the Rust counterpart
- added `StackId` model to the `Document` and `StackIdAdapter` to enable Hive to store this piece of information
- updated tests

### CODE REVIEW

File: discovery_engine/lib/src/domain/document_manager.dart [hunk 1]
```diff
@@ -59,6 +67,16 @@ class DocumentManager {
       throw ArgumentError('id $id does not identify an active document');
     }
     await documentRepo.update(doc..feedback = feedback);
+    final smbertEmbedding = await activeRepo.smbertEmbeddingById(id);
+    if (smbertEmbedding == null) {
+      throw ArgumentError('id $id does not have active data attached');
```
Comment (acrrd) [inline discovery_engine/lib/src/domain/document_manager.dart:72]: This seems more a `StateError`, if the document is active it must have this.

File: discovery_engine/lib/src/domain/document_manager.dart [hunk 2]
```diff
@@ -88,5 +106,16 @@ class DocumentManager {
     }
     activeData.addViewTime(mode, Duration(seconds: sec));
     await activeRepo.update(id, activeData);
+
+    // As we don't have a `DocumentViewMode` on the Rust side at the moment,
+    // we need to decide which value to use or to aggregate all view modes.
+    final sumDuration = activeData.viewTime.values
```
Comment (acrrd) [inline discovery_engine/lib/src/domain/document_manager.dart:112]: Can you ask pink what they want?

File: discovery_engine/lib/src/domain/engine/engine.dart [hunk 3]
```diff
@@ -12,13 +12,33 @@
...
+  void timeLogged(
+    DocumentId docId, {
+    required Uint8List smbertEmbedding,
+    required Duration seconds,
+  });
+
+  /// Process the feedback about the user reacting to a document.
+  void userReacted(
+    DocumentId docId, {
+    // TODO: do we need it, or could the stackId be determined on the rust side?
```
Comment (acrrd) [inline discovery_engine/lib/src/domain/engine/engine.dart:39]: We need it, if is not there already is a data that we need to store in the db

File: discovery_engine/lib/src/domain/document_manager.dart (hunk 2 above)
Comment (biernacki) [reply discovery_engine/lib/src/domain/document_manager.dart:112]: @bartlomiej-matejczy confirmed it should be sum of all view times.

File: discovery_engine/lib/src/domain/engine/engine.dart [hunk 4]
```diff
@@ -12,13 +12,33 @@
...
+    DocumentId docId, {
+    required Uint8List smbertEmbedding,
+    required Duration seconds,
+  });
+
+  /// Process the feedback about the user reacting to a document.
+  void userReacted(
+    DocumentId docId, {
+    required StackId stackId,
+    required String snippet,
```
Comment (biernacki) [inline discovery_engine/lib/src/domain/engine/engine.dart:40]: snippet was added cause @Robert-Steiner is doing same in the Rust counterpart

Comment (Robert-Steiner) [reply discovery_engine/lib/src/domain/engine/engine.dart:40]: #72 has been merged

File: discovery_engine/lib/src/domain/models/document.dart [hunk 5]
```diff
@@ -31,21 +31,24 @@ class Document {
   @HiveField(0)
   final DocumentId documentId;
   @HiveField(1)
-  final WebResource webResource;
+  final StackId stackId;
```
Comment (acrrd) [inline discovery_engine/lib/src/domain/models/document.dart:34]: Now this is ok but in the future we will not be able to shuffle the field ids.

File: discovery_engine/lib/src/infrastructure/type_adapters/hive_unique_id_adapter.dart [hunk 6]
```diff
@@ -14,23 +14,37 @@
...
   }
+}
 
+class DocumentIdAdapter extends _UniqueIdAdapter<DocumentId> {
   @override
-  void write(BinaryWriter writer, DocumentId obj) {
-    final bytes = obj.value.buffer.asUint8List();
-    writer.writeByteList(bytes);
-  }
+  final typeId = documentIdTypeId;
```
Comment (acrrd) [inline discovery_engine/lib/src/infrastructure/type_adapters/hive_unique_id_adapter.dart:44]: I'm surprised that this does not trigger https://dart-lang.github.io/linter/lints/overridden_fields.html

**Review by acrrd — APPROVED** [review]

File: discovery_engine/lib/src/domain/models/document.dart (hunk 5 above)
Comment (biernacki) [reply discovery_engine/lib/src/domain/models/document.dart:34]: yes, I'm aware, but I thought it's ok, since we didn't release the app anywhere yet

Comment (acrrd) [reply discovery_engine/lib/src/domain/models/document.dart:34]: Yep now is ok

File: discovery_engine/lib/src/infrastructure/type_adapters/hive_unique_id_adapter.dart (hunk 6 above)
Comment (biernacki) [reply discovery_engine/lib/src/infrastructure/type_adapters/hive_unique_id_adapter.dart:44]: but it's not overriding a field, it's assigning it, as the base class defines a getter
```dart
abstract class TypeAdapter<T> {
  /// Called for type registration
  int get typeId;
}
```

**Pull request merged**